torch/onnx/utils.py

```diff
@@ -1660,6 +1625,110 @@ def _export(
+        _exporter_states.ExportTypes.COMPRESSED_ZIP_ARCHIVE,
+    ]:
+        compression = (
+            zipfile.ZIP_DEFLATED
+            if export_type == _exporter_states.ExportTypes.COMPRESSED_ZIP_ARCHIVE
+            else zipfile.ZIP_STORED
+        )
+        with zipfile.ZipFile(f, "w", compression=compression) as z:
+            z.writestr(_constants.ONNX_ARCHIVE_MODEL_PROTO_NAME, model_bytes)
+            for k, v in export_map.items():
+                z.writestr(k, v)
+    elif export_type == _exporter_states.ExportTypes.DIRECTORY:
+        if os.path.exists(f):  # type: ignore[arg-type]
+            assert os.path.isdir(f)  # type: ignore[arg-type]
+        else:
+            os.makedirs(f)  # type: ignore[arg-type]
+
+        model_proto_file = os.path.join(f, _constants.ONNX_ARCHIVE_MODEL_PROTO_NAME)  # type: ignore[arg-type]
+        with torch.serialization._open_file_like(model_proto_file, "wb") as opened_file:
+            opened_file.write(model_bytes)
+
+        for k, v in export_map.items():
+            weight_proto_file = os.path.join(f, k)  # type: ignore[arg-type]
+            with torch.serialization._open_file_like(
+                weight_proto_file, "wb"
+            ) as opened_file:
+                opened_file.write(v)
+    else:
+        raise RuntimeError("Unknown export type")
+
+
+@_beartype.beartype
+def _add_onnxscript_fn(
+    model_bytes: bytes,
+    custom_opsets: Mapping[str, int],
+) -> bytes:
+    """Insert model-included custom onnx-script function into ModelProto"""
+
+    # TODO(titaiwang): remove this when onnx becomes dependency
+    try:
+        import onnx
+    except ImportError:
+        raise errors.OnnxExporterError("Module onnx is not installed!")
+
+    # For > 2GB model, onnx.load_fromstring would fail. However, because
+    # in _export_onnx, the tensors should be saved separately if the proto
+    # size > 2GB, and if it for some reason did not, the model would fail on
+    # serialization anyway in terms of the protobuf limitation. So we don't
+    # need to worry about > 2GB model getting here.
+    model_proto = onnx.load_from_string(model_bytes)
+
+    # Iterate graph nodes to insert only the included custom
+    # function_proto into model_proto
+    # TODO(titaiwang): Currently, onnxscript doesn't support ONNXFunction
+    # calling other ONNXFunction scenario, neither does it here
+    onnx_function_list = list()
+    included_node_func = set()
+    for node in model_proto.graph.node:
+        node_kind = node.domain + "::" + node.op_type
+        if (
+            jit_utils.is_custom_domain(node.domain)
+            and node_kind not in included_node_func
+        ):
+            specified_version = custom_opsets.get(node.domain, 1)
+            onnx_function_group = registration.registry.get_function_group(node_kind)
+            if onnx_function_group is not None:
+                onnx_fn = onnx_function_group.get(specified_version)
```

Comment:
One type of likely user error is the version of `onnx_fn` they passed in `onnxscript_op` doesn't match the `specified_version`.

Reply:
And how to ensure the onnx version in `onnx_fn` is compatible with that in `model_proto`?

Reply:
So we have two opset versions here. One is custom op version, the other is opset_version of model_proto.
Let's say we register a `Selu` with custom opset version 1, and implemented with opset 15. In this case, If we call `torch.onnx.export` with opset_version < 15, this custom op would not be consumed in the model_proto, based on the current registry design and my experiment. In another case, if user wrongly call `torch.onnx.export` with custom op version 2, then he/she would fallback to custom opset version 1, based on the logic here.
LMK if I anwser this question?

Reply:
Thanks, that makes sense. I looked into #86907 to understand 
> If we call torch.onnx.export with opset_version < 15, this custom op would not be consumed in the model_proto